---
title: "HNSCC Proteomics Plotting"
author: "Sara Gosline, Song Feng"
date: "6/30/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(grid)
library(gridExtra)
```

## Summary
The goal of this analysis is to ascertain the expression of key proteins in the HSNCC dataset that could give rise to imaging features.

```{r load data, echo=FALSE,warning=FALSE}
library(readxl)
library(dplyr)
library(tidyr)
##read in imaging features 
imgFeat <-readxl::read_xlsx('../data/cptac_hnc_pyrad_features_v2.xlsx')
##convert to long format
imgFeat <- imgFeat %>%
  pivot_longer(3:ncol(imgFeat),names_to='patient',values_to='value')

##read in proteomics
impProts <- read.table('../data/S054_HNSCC_imputed_0920.tsv',sep='\t',header=T,check.names=F)

##convert to long format
tumProts <- impProts%>%
  dplyr::select(Gene,ends_with('-T')) 

##convert to long format
normProts <- impProts%>%
  dplyr::select(Gene,ends_with('-N'))

normProts <- normProts%>%
  pivot_longer(2:ncol(normProts),names_to='patient',values_to='inferredNormAbund')%>%
  mutate(patient=stringr::str_remove(patient,'-N'))

tumProts <- tumProts%>%
  pivot_longer(2:ncol(tumProts),names_to='patient',values_to='inferredTumAbund')%>%
  mutate(patient=stringr::str_remove(patient,'-T'))


##can we get lcinical too?
clin.dat<-read.table('../data/HNSCC_Feb2021_clinical_data.csv',sep=',',header=T,check.names=F,quote='"')%>%
  dplyr::select(case_id,contains("lymph_node"))%>%### this gets us 8 different categories
  rename(extranodal_extension="baseline/lymph_nodes_extranodal_extension",
         staging="baseline/pathologic_staging_regional_lymph_nodes_pn",
         dissection_performed="baseline/lymph_node_neck_dissection_performed",
         dissection_left="baseline/method_of_lymph_node_neck_dissection_left",
         dissection_right="baseline/method_of_lymph_node_neck_dissection_right",
         ln_examined="baseline/number_of_lymph_nodes_examined",
        positive_by_he="baseline/number_of_lymph_nodes_positive_for_tumor_by_he",
        postiive_by_ihc="baseline/number_of_lymph_nodes_positive_for_tumor_by_ihc_staining_only",
        patient='case_id')%>%
  mutate(frac_ln=positive_by_he/ln_examined)

##now combine the abundances to the same table
fullProts <- normProts%>%inner_join(tumProts,by=c('Gene','patient'))%>%
  mutate(tumNormDiff=inferredTumAbund/inferredNormAbund)%>%
  left_join(clin.dat,by='patient')


## read in gene list
invList <-read.table('../data/pouliquenEtAlProteins.tsv',sep='\t',header=T,check.names=F)

res=data.frame(`Stats`=c('total proteins','proteins in list','total samples','total with normal and tumor'),
           `Number`=c(length(unique(impProts$Gene)),length(intersect(invList$Symbol,fullProts$Gene)),
                       ncol(impProts)-1,length(unique(fullProts$patient))))

print(res)
```
Now that we have the genes and the samples aligned we can begin to analyze them more broadly.


## Visualize samples by patient clustered by tumor abundance

We now want to see how the patients fall across the various variables. We start by plotting raw tumor abundance, then compute differences between tumor/normal.
```{r pcaplot, echo=FALSE}


##quick function to do PCA on a column value
pcaFromDf<-function(df,column){
  mat <-df%>%
    dplyr::select('Gene','patient',column)%>%
    pivot_wider(names_from='patient',values_from=column)%>%
    tibble::column_to_rownames('Gene')%>%
    as.matrix()
  return(data.frame(prcomp(t(mat))$x)%>%tibble::rownames_to_column('patient'))
}

##let's first do a PCA plotted with positive lns
tumAbund <- pcaFromDf(fullProts,'inferredTumAbund')%>%
  left_join(clin.dat,by='patient')

##all proteins
ggplot(tumAbund)+
  geom_point(aes(x=PC1,y=PC2,col=ln_examined,size=frac_ln))+
  ggtitle('All proteins tumor expression')

##let's first do a PCA plotted with positive lns
tumNorm <- pcaFromDf(fullProts,'tumNormDiff')%>%
  left_join(clin.dat,by='patient')

##all proteins
ggplot(tumNorm)+
  geom_point(aes(x=PC1,y=PC2,col=ln_examined,size=frac_ln))+
  ggtitle('Tumor vs. normal all proteins')

```


## Selected invasion proteins
Do we see any difference  in clustering/staining with invasivity?

```{r pcaplot2, echo=FALSE}

redProts <- fullProts%>%
  subset(Gene%in%invList$Symbol)

##let's first do a PCA plotted with positive lns
tumAbund <- pcaFromDf(redProts,'inferredTumAbund')%>%
  left_join(clin.dat,by='patient')

##all proteins
ggplot(tumAbund)+
  geom_point(aes(x=PC1,y=PC2,col=ln_examined,size=frac_ln))+
  ggtitle('Invasive proteins tumor expression')

##let's first do a PCA plotted with positive lns
tumNorm <- pcaFromDf(redProts,'tumNormDiff')%>%
  left_join(clin.dat,by='patient')

##all proteins
ggplot(tumNorm)+
  geom_point(aes(x=PC1,y=PC2,col=ln_examined,size=frac_ln))+
  ggtitle('Tumor vs. normal invasive proteins')

```

```{r}
library(reshape)
```


```{r}
length(tumProts$patient)
length(unique(tumProts$patient))
length(normProts$patient)
length(unique(normProts$patient))
length(imgFeat$patient)
length(unique(imgFeat$patient))
```

```{r}
tumProtTable <- cast(tumProts, patient~Gene, mean)
rownames(tumProtTable) <- tumProtTable$patient
tumProtTable$patient <- NULL
normProtTable <- cast(normProts, patient~Gene, mean)
rownames(normProtTable) <- normProtTable$patient
normProtTable$patient <- NULL
imgFeatTable <- cast(imgFeat, patient~Feature_Type, mean)
rownames(imgFeatTable) <- imgFeatTable$patient
imgFeatTable$patient <- NULL
```

```{r}
# if(!require(devtools)) install.packages("devtools")
#   devtools::install_github("kassambara/ggcorrplot")
#install.packages("ggcorrplot")
library(ggcorrplot)

crossCorr <- function(table1, table2) {
  #colNames <- Reduce(intersect, list(colnames(table1), colnames(table2)))
  rowNames <- Reduce(intersect, list(rownames(table1), rownames(table2)))
  t1 <- as.data.frame(table1[rowNames,])
  t2 <- as.data.frame(table2[rowNames,])
  
  return(cor(t1, t2))
}

```

```{r}
corrIF <- cor(imgFeatTable)
ggcorrplot(corrIF)
ggsave("corr_imgFeat.pdf", width = 25, height = 20)
```

```{r}
corrIfTp <- crossCorr(imgFeatTable, tumProtTable)
ggcorrplot(corrIfTp)
ggsave("corr_imgFeat_tumProt.pdf", width = 10, height = 100, limitsize = FALSE)
```


```{r}
corrIfNp <- crossCorr(imgFeatTable, normProtTable)
ggcorrplot(corrIfNp)
ggsave("corr_imgFeat_normProt.pdf", width = 10, height = 100, limitsize = FALSE)
```

```{r}
corrDiff <- corrIfTp - corrIfNp
ggcorrplot(corrDiff)
ggsave("corrDiff_imgFeat_tumProt_vs_imgFeat_normProt.pdf", width = 10, height = 100, limitsize = FALSE)
```

```{r}
save(corrIfTp, corrIfNp, corrDiff, corrIF, file = "allRawCorrs.RData")
```

# TODOs

0. Filter the correlation results, get some features out first.

1. Get a ranked list based on correlation resutls, and do enrichment analysis
  1.1. LeapR: https://github.com/biodataganache/leapR
  1.2. https://github.com/PNNL-CompBio/earlyLateAMLresistance/blob/main/stromaClinicalTrial/global/12-GSEA_and_correlation_enrichment.Rmd
  1.3. https://github.com/sgosline/tdlnRadProt/blob/main/prelimPredModels.py
2. Send the enrichment analysis results to Sara



The next step might be picking these highly correlated features and genes to build a bipartite graph.

```{r}
filterCorr <- function(corTable, th = 0.6) {
  return(as.data.frame(corTable[,colSums((corTable > th) | (corTable < -th))> 0]))
}
```

```{r}
corrIfTpF <- filterCorr(corrIfTp)
print(dim(corrIfTpF))
ggcorrplot(corrIfTpF)
ggsave("corrF_imgFeat_tumProt.pdf", width = 25, height = 8, limitsize = FALSE)
```



```{r}
corrIfTpF <- filterCorr(corrIfTp,0.5)
print(dim(corrIfTpF))
ggcorrplot(corrIfTpF)
ggsave("corrF_imgFeat_tumProt_0.5.pdf", width = 25, height = 30, limitsize = FALSE)
```




```{r}
corrIfTpF <- filterCorr(corrIfTp,0.7)
print(dim(corrIfTpF))
ggcorrplot(corrIfTpF)
ggsave("corrF_imgFeat_tumProt_0.7.pdf", width = 25, height = 8, limitsize = FALSE)
```


```{r}
corrIfNpF <- filterCorr(corrIfNp)
print(dim(corrIfNpF))
ggcorrplot(corrIfNpF)
ggsave("corrF_imgFeat_normProt.pdf", width = 25, height = 30, limitsize = FALSE)
```


```{r}
corrIfNpF <- filterCorr(corrIfNp, 0.7)
print(dim(corrIfNpF))
ggcorrplot(corrIfNpF)
ggsave("corrF_imgFeat_normProt_0.7.pdf", width = 25, height = 8, limitsize = FALSE)
```

```{r}
corrDiffF <- filterCorr(corrDiff, 0.9)
print(dim(corrDiffF))
ggcorrplot(corrDiffF)
ggsave("corrDiffF_imgFeat_tumProt_vs_imgFeat_normProt_0.9.pdf", width = 25, height = 10, limitsize = FALSE)
```



```{r}
save(corrIfTpF, corrIfNpF, corrDiffF, file = "allCorrs.RData")

write.table(corrDiffF, file = "corrDiffF.tsv", quote = F, sep = '\t', col.names=NA)
write.table(corrIfTpF, file = "corrIfTpF.tsv", quote = F, sep = '\t', col.names=NA)
write.table(corrIfNpF, file = "corrIfNpF.tsv", quote = F, sep = '\t', col.names=NA)

write.table(t(corrDiffF), file = "corrDiffFT.tsv", quote = F, sep = '\t', col.names=NA)
write.table(t(corrIfTpF), file = "corrIfTpFT.tsv", quote = F, sep = '\t', col.names=NA)
write.table(t(corrIfNpF), file = "corrIfNpFT.tsv", quote = F, sep = '\t', col.names=NA)
```


```{r}
load("allRawCorrs.RData")
load("allCorrs.RData")
```


```{r}

```



```{r}
library(leapr)
```

```{r}
print(rownames(corrIfNpF))
```


```{r}
features <- c("Complexity", "InterquartileRange", "LeastAxisLength", "MajorAxisLength", "Maximum2DDiameterRow", "Maximum2DDiameterSlice", "MeshVolume", "SizeZoneNonUniformity", "SmallDependenceHighGrayLevelEmphasis", "SurfaceArea", "TotalEnergy", "VoxelVolume")
corrMatrix <- corrIfNpF
print(sapply(features, function(x) length(corrMatrix[x,][corrMatrix[x,] > 0.6 | corrMatrix[x,] < -0.6])))
# z.en <- leapR(geneset=ncipid,enrichment_method='enrichment_in_order',datamatrix=oneVsZero,primary_columns='logFC') %>% subset(BH_pvalue<0.05)
# print(z.en)
```

```{r}
thr <- 0.6
corrMatrix <- filterCorr(corrIfTp[features,], thr)
print(corrMatrix)
print(sapply(features, function(x) length(corrMatrix[x,][corrMatrix[x,] > thr | corrMatrix[x,] < -thr])))
ggcorrplot(corrMatrix)
ggsave(paste0("corr_selected_imgFeat_tumProt_", thr, ".pdf"), width = 6, height = 6, limitsize = FALSE)
```


```{r}
thr <- 0.6
corrMatrix <- filterCorr(corrIfNp[features,], thr)
print(corrMatrix)
print(sapply(features, function(x) length(corrMatrix[x,][corrMatrix[x,] > thr | corrMatrix[x,] < -thr])))
ggcorrplot(corrMatrix)
ggsave(paste0("corr_selected_imgFeat_normProt_", thr, ".pdf"), width = 6, height = 6, limitsize = FALSE)
```

```{r}
thr <- 0.5
corrMatrix <- filterCorr(corrIfTp[features,], thr)
print(corrMatrix)
print(sapply(features, function(x) length(corrMatrix[x,][corrMatrix[x,] > thr | corrMatrix[x,] < -thr])))
ggcorrplot(corrMatrix)
ggsave(paste0("corr_selected_imgFeat_tumProt_", thr, ".pdf"), width = 6, height = 6, limitsize = FALSE)
```

```{r}
thr <- 0.5
corrMatrix <- filterCorr(corrIfNp[features,], thr)
print(corrMatrix)
print(sapply(features, function(x) length(corrMatrix[x,][corrMatrix[x,] > thr | corrMatrix[x,] < -thr])))
ggcorrplot(corrMatrix)
ggsave(paste0("corr_selected_imgFeat_normProt_", thr, ".pdf"), width = 6, height = 30, limitsize = FALSE)
```

```{r}
corrNpSlIf <- t(corrIfNp[features,])
corrTpSlIf <- t(corrIfTp[features,])
```

```{r}
z.en <- leapR(geneset=ncipid,enrichment_method='enrichment_in_order',datamatrix=corrNpSlIf,primary_columns='TotalEnergy')
min(z.en$BH_pvalue, na.rm = T)
print(z.en %>% subset(BH_pvalue<0.1))
```


```{r}
corrMatrix <- corrTpSlIf
for (feature in features) {
  z.en <- leapR(geneset=ncipid,enrichment_method='enrichment_in_order',datamatrix=corrMatrix,primary_columns=feature) %>% subset(BH_pvalue<0.05)
  print(z.en)
  write.table(z.en, file = paste0("enrichment/", feature, "-tumProt-enrichment.tsv"), quote = F, sep = '\t', col.names=NA)
}
```


```{r}
corrMatrix <- corrNpSlIf
for (feature in features) {
  z.en <- leapR(geneset=ncipid,enrichment_method='enrichment_in_order',datamatrix=corrMatrix,primary_columns=feature) %>% subset(BH_pvalue<0.05)
  print(z.en)
  write.table(z.en, file = paste0("enrichment/", feature, "-normProt-enrichment.tsv"), quote = F, sep = '\t', col.names=NA)
}
```



```{r}
corrMatrix <- abs(corrTpSlIf)
for (feature in features) {
  z.en <- leapR(geneset=ncipid,enrichment_method='enrichment_in_order',datamatrix=corrMatrix,primary_columns=feature) %>% subset(BH_pvalue<0.05)
  print(z.en)
  write.table(z.en, file = paste0("enrichment/", feature, "-tumProt-enrichment-abs.tsv"), quote = F, sep = '\t', col.names=NA)
}
```



```{r}
corrMatrix <- abs(corrNpSlIf)
for (feature in features) {
  z.en <- leapR(geneset=ncipid,enrichment_method='enrichment_in_order',datamatrix=corrMatrix,primary_columns=feature) %>% subset(BH_pvalue<0.05)
  print(z.en)
  write.table(z.en, file = paste0("enrichment/", feature, "-normProt-enrichment-abs.tsv"), quote = F, sep = '\t', col.names=NA)
}
```

```{r}

```




