---
title: "HNSCC Proteomics Plotting"
author: "Sara Gosline"
date: "6/30/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(grid)
library(gridExtra)
```

## Summary
The goal of this analysis is to ascertain the expression of key proteins in the HSNCC dataset that could give rise to imaging features.

```{r load data, echo=FALSE,warning=FALSE}
source("loadHNSCCdata.R")

res=data.frame(`Stats`=c('total proteins','proteins in list','total samples','total with normal and tumor'),
           `Number`=c(length(unique(impProts$Gene)),length(intersect(invList$Symbol,fullProts$Gene)),
                       ncol(impProts)-1,length(unique(fullProts$patient))))

print(res)
```
Now that we have the genes and the samples aligned we can begin to analyze them more broadly.


## Visualize samples by patient clustered by tumor abundance

We now want to see how the patients fall across the various variables. We start by plotting raw tumor abundance, raw normal abundance, then compute differences between tumor/normal.

```{r pcaplot, warning=FALSE}

##quick function to do PCA on a column value
pcaFromDf<-function(df,column){
  mat <-df%>%
    dplyr::select('Gene','patient',column)%>%
    pivot_wider(names_from='patient',values_from=column)%>%
    tibble::column_to_rownames('Gene')%>%
    as.matrix()
  return(data.frame(prcomp(t(mat))$x)%>%tibble::rownames_to_column('patient'))
}

##let's first do a PCA plotted with positive lns
tumAbund <- pcaFromDf(fullProts,'inferredTumAbund')%>%
  left_join(clin.dat,by='patient')%>%
  left_join(select(red.dat,c(patient,lnCounts)),by='patient')


##all proteins
ggplot(tumAbund)+
  geom_point(aes(x=PC1,y=PC2,size=ln_examined,col=lnCounts))+
  ggtitle('All proteins tumor expression')


normAbund <- pcaFromDf(normProts,'inferredNormAbund')%>%
  left_join(clin.dat,by='patient')%>%
  left_join(select(red.dat,c(patient,lnCounts)),by='patient')

##now let's plot ll proteins normal expression
ggplot(normAbund)+
  geom_point(aes(x=PC1,y=PC2,size=ln_examined,col=lnCounts))+
  ggtitle('All proteins normal expression')

##let's first do a PCA plotted with positive lns
tumNorm <- pcaFromDf(fullProts,'tumNormDiff')%>%
  left_join(clin.dat,by='patient')%>%
  left_join(select(red.dat,c(patient,lnCounts)),by='patient')


##all proteins
ggplot(tumNorm)+
  geom_point(aes(x=PC1,y=PC2,col=lnCounts,size=ln_examined))+
  ggtitle('Tumor vs. normal all proteins')

```


## Selected invasion proteins
We have a set of proteins that are known to be related to invasivity - do they cluster the samples any better? 

Do we see any difference  in clustering/staining with invasivity

```{r pcaplot2, echo=FALSE}

redProts <- fullProts%>%
  subset(Gene%in%invList$Symbol)

##let's first do a PCA plotted with positive lns
tumAbund <- pcaFromDf(redProts,'inferredTumAbund')%>%
  left_join(clin.dat,by='patient')%>%left_join(select(red.dat,c(patient,lnCounts)),by='patient')

##all proteins
ggplot(tumAbund)+
  geom_point(aes(x=PC1,y=PC2,col=lnCounts,size=frac_ln))+
  ggtitle('Invasive proteins tumor expression')

normAbund <-pcaFromDf(redProts,'inferredNormAbund')%>%
    left_join(clin.dat,by='patient')%>%left_join(select(red.dat,c(patient,lnCounts)),by='patient')

ggplot(normAbund)+
  geom_point(aes(x=PC1,y=PC2,col=lnCounts,size=frac_ln))+
  ggtitle('Invasive proteins normal expression')


##let's first do a PCA plotted with positive lns
tumNorm <- pcaFromDf(redProts,'tumNormDiff')%>%
  left_join(clin.dat,by='patient')%>%
  left_join(select(red.dat,c(patient,lnCounts)),by='patient')


##all proteins
ggplot(tumNorm)+
  geom_point(aes(x=PC1,y=PC2,col=lnCounts,size=frac_ln))+
  ggtitle('Tumor vs. normal invasive proteins')

```


## Conclusion
There is no clear separation between samples based on all proteins or invasive proteins.